main_boaz.tex

```diff
@@ -0,0 +1,910 @@
+
+
+  \item An element of an axis $\name{ax}$ is known as an \emph{index}. We think of an axis $\name{name}$ as describing a \emph{type} that inherits from the type $X$. Therefore, if (for example) $Supp(\name{ax})$ is the integers then we can use operations such as addition, subtraction, multiplication on indices.  Even if two axes have the same support, we still require explicit casting to translate between one to the other, and so (for example) if $i \in \name{batch}$ then it cannot be used as is to axis into a vector of shape $\{ \name{time} \}$. In contrast, if two axes differ only in annotation, then we can translate indices between one and the other without explicit casting. We can think of an annotated axis of the form $\name{ax}_annot$ as a parameter of type $\name{ax}$ and name \texttt{ax\_annot}. (In this sense annotation is like ``Hungarian notation'' for parameter names.) 
+
+
+  
+  \item A \emph{slice} is a subset of an axis. (This does not need to be a strict subset, and so the ax itself is also a valid slice.)  If \name{ax} is an axis and $S \subseteq Supp(\name{ax})$ then  $\name{ax}[S]$ is the slice corresponding to all triples $(ax,annot,x) \in \name{ax}$ such that $x\in S$.  If $Supp(\name{ax})$ is a set of integers, then we use the notation $\name{ax}[..n]$ for the $\name{ax}[\{1,\ldots, n \}]$. We will sometimes use annotation and slicing together and so use the notation such as $\name{layer}_1 = \name{layer}[..128]$ to denote that $\name{layer}_1$ is the annotation of $\name{layer}$ with support $\{1,..,128 \}$.  
+   
+  \item A (tensor) \emph{type} $\mathcal{T}$ is an unordered set of axes. If $\mathcal{T} = \{ \name{ax}_1 ,\ ldots, \name{ax}_d \}$ is a type then a \emph{shape} of type $\mathcal{T}$ is an unordered set of the form $\{ \name{ax}_1[S_1] ,\ldots, \name{ax}_d[S_d] \}$. That is, a shape of type $\mathcal{T}$ is a set of slices of the axes of $\mathcal{T}$. For example, $\{ \name{width} , \name{height}, \name{batch} \}$ is a tensor type while $\{ \name{width}[32], \name{height}[32], \name{batch}[128] \}$ is a shape. Note in both shapes and types, there are no two axes or slices that have both identical names and annotations.
+
+  \item For every shape $\mathcal{S} = \{ S_1,\ldots, S_d \}$ (where the $S_i$'s are slices, we denote by $ind \mathcal{S}$ the set $\{ \{ i_1,\ldots, i_d \} | i_1 \in S_1, \ldots, i_d \in S_d \}$. An element of $ind \mathcal{S}$ is called a \emph{coordinate} (since it specified a unique point in all the axes). A (named) \emph{tensor} with shape $\mathcal{S}$ is a map from $ind \mathcal{S}$ to $\mathbb{R}$. We denote by $\mathbb{R}^{\mathcal{S}}$ to be the set of tensors with shape $\mathcal{S}$ and $\mathbb{R}^{\mathcal{T}}$ to be the union of $\mathbb{R}^{\mathcal{S}}$ for all shapes $\mathcal{S}$ of type $\mathcal{T}$.
```

Comment:
If S has two axes ax^annot1 and ax^annot2, there's a little bit of ambiguity here whether, in ind(S), the coordinates {(ax,annot1,x),(ax,annot2,y)} and {(ax,annot1,y),(ax,annot2,x)} would be considered distinct coordinates, since we allow automatic conversion between indices with different annotations.

Reply:
They are considered distinct as coordinates (since that's what annotations are for) but we can cast between them automatically, so for example we don't need to say anything special if we take a dot product of a {  **layer**_1 } and { **layer**_2 } shaped vectors. We only worry about the annotation when a tensor has two different axes with the same name (and then indeed it matters).